controllers/clusterpermission_controller.go

```diff
@@ -527,3 +541,159 @@ func (r *ClusterPermissionReconciler) generateManifestWorkPayload(ctx context.Co
 
 	return clusterRole, clusterRoleBindings, roles, roleBindings, nil
 }
+
+// handleValidation processes role validation when spec.validate is enabled
+func (r *ClusterPermissionReconciler) handleValidation(ctx context.Context, clusterPermission *cpv1alpha1.ClusterPermission) error {
+	log := log.FromContext(ctx)
+
+	// Extract role references that need validation
+	roleRefs := extractRoleReferencesForValidation(clusterPermission)
+	if len(roleRefs) == 0 {
+		log.Info("no role references found for validation")
+		return nil
+	}
+
+	validationMWName := generateValidationManifestWorkName(*clusterPermission)
+	validationManifestWork := buildValidationManifestWork(*clusterPermission, validationMWName, roleRefs)
+
+	// Create or update validation ManifestWork
+	var existingValidationMW workv1.ManifestWork
+	err := r.Get(ctx, types.NamespacedName{Name: validationMWName, Namespace: clusterPermission.Namespace}, &existingValidationMW)
+	if apierrors.IsNotFound(err) {
+		log.Info("creating validation ManifestWork")
+		err = r.Client.Create(ctx, validationManifestWork)
+		if err != nil {
+			log.Error(err, "unable to create validation ManifestWork")
+			return err
+		}
+	} else if err == nil {
+		log.Info("updating validation ManifestWork")
+		existingValidationMW.Spec = validationManifestWork.Spec
+		err = r.Client.Update(ctx, &existingValidationMW)
+		if err != nil {
+			log.Error(err, "unable to update validation ManifestWork")
+			return err
+		}
+	} else {
+		log.Error(err, "unable to fetch validation ManifestWork")
+		return err
+	}
+
+	// Process validation results from ManifestWork status
+	return r.processValidationResults(ctx, clusterPermission, validationMWName)
+}
+
+// processValidationResults processes the feedback from validation ManifestWork and updates status conditions
+func (r *ClusterPermissionReconciler) processValidationResults(ctx context.Context, clusterPermission *cpv1alpha1.ClusterPermission, validationMWName string) error {
+	log := log.FromContext(ctx)
+
+	// Get the validation ManifestWork to check its status
+	var validationMW workv1.ManifestWork
+	success := false
+	for i := 0; i < VALIDATION_MW_RETRY_COUNT; i++ {
+		log.Info("checking validation ManifestWork status", "attempt", i+1)
+		err := r.Get(ctx, types.NamespacedName{Name: validationMWName, Namespace: clusterPermission.Namespace}, &validationMW)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				// ManifestWork doesn't exist yet, nothing to process
+				return nil
+			}
+			return err
+		}
+		if validationMW.Status.Conditions[0].Type == "Applied" && validationMW.Status.Conditions[0].Status == "True" {
```

Comment:
This is not the right way to fetch condition because there is no guarantee the first element is Applied. Refer to the clustercurrator where we used something like meta.findconditiontype

Reply:
Fixed.